servers/src/main/java/tachyon/master/file/FileSystemMaster.java

```diff
@@ -1478,8 +1517,182 @@ public void setState(long fileId, SetStateOptions options)
     }
   }
 
-  private void setStateInternal(long fileId, long opTimeMs, SetStateOptions options)
+  /**
+   * Schedules a file for async persistence.
+   *
+   * @param fileId the id of the file for persistence
+   * @return the id of the worker that persistence is scheduled on
+   * @throws FileDoesNotExistException when the file does not exist
+   */
+  public long scheduleAsyncPersistence(long fileId) throws FileDoesNotExistException {
+    // find the worker
+    long workerId = getWorkerStoringFile(fileId);
+
+    if (workerId == IdUtils.INVALID_WORKER_ID) {
+      LOG.warn("No worker found to schedule async persistence for file {}", fileId);
+      // no worker found, do nothing
+      return workerId;
+    }
+
+    // update the state
+    synchronized (mInodeTree) {
+      Inode inode = mInodeTree.getInodeById(fileId);
+      inode.setPersistenceState(PersistenceState.IN_PROGRESS);
+    }
+
+    synchronized (mWorkerToAsyncPersistFiles) {
+      if (!mWorkerToAsyncPersistFiles.containsKey(workerId)) {
+        mWorkerToAsyncPersistFiles.put(workerId, Sets.<Long>newHashSet());
+      }
+      mWorkerToAsyncPersistFiles.get(workerId).add(fileId);
+    }
+
+    // TODO(yupeng) TACHYON-1456 add fault tolerance and flush journal
+
+    return workerId;
+  }
+
+  /**
+   * Gets a worker where the given file is stored.
+   *
+   * @param fileId the file id, -1 if no worker can be found
+   * @return the storing worker
+   * @throws FileDoesNotExistException when the file does not exist on any worker
+   */
+  private long getWorkerStoringFile(long fileId) throws FileDoesNotExistException {
+    Map<Long, Integer> workerBlockCounts = Maps.newHashMap();
+    List<FileBlockInfo> blockInfoList;
+    try {
+      blockInfoList = getFileBlockInfoList(fileId);
+
+      for (FileBlockInfo fileBlockInfo : blockInfoList) {
+        for (BlockLocation blockLocation : fileBlockInfo.blockInfo.locations) {
+          if (workerBlockCounts.containsKey(blockLocation.workerId)) {
+            workerBlockCounts.put(blockLocation.workerId,
+                workerBlockCounts.get(blockLocation.workerId) + 1);
+          } else {
+            workerBlockCounts.put(blockLocation.workerId, 1);
+          }
+        }
+      }
+    } catch (FileDoesNotExistException e) {
+      LOG.error("The file {} to persist does not exist", fileId);
+      return IdUtils.INVALID_WORKER_ID;
+    } catch (InvalidPathException e) {
+      LOG.error("The file {} to persist does not exist", fileId);
+      return IdUtils.INVALID_WORKER_ID;
+    }
+
+    if (workerBlockCounts.size() == 0) {
+      LOG.error("The file " + fileId + " does not exist on any worker");
```

Comment:
Is this recoverable or will an exception be thrown later? `LOG.warn` is generally used if the issue is bad but an expected case, whereas `error` implies something will be broken (ie. exception thrown/program exit)